packages/flutter/lib/src/cupertino/nav_bar.dart

```diff
@@ -809,58 +809,73 @@ class _LargeTitleNavigationBarSliverDelegate
-                          child: DefaultTextStyle(
-                            style: CupertinoTheme.of(context).textTheme.navLargeTitleTextStyle,
-                            maxLines: 1,
-                            overflow: TextOverflow.ellipsis,
-                            child: components.largeTitle!,
+      // `LayoutBuilder` provides the height of the nav bar for the large
+      // title magnification.
+      child: LayoutBuilder(
+        builder: (BuildContext context, BoxConstraints constraints) {
+          return DefaultTextStyle(
+            style: CupertinoTheme.of(context).textTheme.textStyle,
+            child: Stack(
+              fit: StackFit.expand,
+              children: <Widget>[
+                Positioned(
+                  top: persistentHeight,
+                  left: 0.0,
+                  right: 0.0,
+                  bottom: 0.0,
+                  child: ClipRect(
+                    // The large title starts at the persistent bar.
+                    // It's aligned with the bottom of the sliver and expands clipped
+                    // and behind the persistent bar.
+                    child: OverflowBox(
+                      minHeight: 0.0,
+                      maxHeight: double.infinity,
+                      alignment: AlignmentDirectional.bottomStart,
+                      child: Padding(
+                        padding: const EdgeInsetsDirectional.only(
+                          start: _kNavBarEdgePadding,
+                          bottom: 8.0, // Bottom has a different padding.
+                        ),
+                        child: SafeArea(
+                          top: false,
+                          bottom: false,
+                          child: AnimatedOpacity(
+                            opacity: showLargeTitle ? 1.0 : 0.0,
+                            duration: _kNavBarTitleFadeDuration,
+                            child: Semantics(
+                              header: true,
+                              child: DefaultTextStyle(
+                                style: CupertinoTheme.of(context).textTheme.navLargeTitleTextStyle,
+                                maxLines: 1,
+                                overflow: TextOverflow.ellipsis,
+                                child: Transform.scale(
+                                  // This scale is estimated from the settings app in iOS 14.
+                                  // The large title scales linearly from 1.0 up to 1.15 magnification.
+                                  // The `constraints.maxHeight` value is the height of the nav bar, 
+                                  // and `maxExtent` is the default large title height the nav bar snaps back to. 
+                                  // The difference between the two heights is used to scale the title.
+                                  scale: math.max(1.0, math.min(1.15, 1 + (constraints.maxHeight - maxExtent) / maxExtent * 0.12)),
```

Comment:
super nit: the logic is probably easier to see with `clamp`: `(1.0 + (constraints.maxHeight - maxExtent) / maxExtent * 0.12)).clamp(1.0, 1.15)`,

Reply:
Is the `maxExtent` here supposed to be `minExtent`? When `constraints.maxHeight == minExtent` the expression should evaluate to 1.0 from the looks of it?

Reply:
Will this have hit testing problems when we scale the title? `Transform.scale` doesn't seem to change the layout size of the render object, so I assume when you change the `scale`, the maximum hit testing region is still the size of the render object:
https://github.com/flutter/flutter/blob/8df56056b4914060784e68e6b149a30319b75d3d/packages/flutter/lib/src/rendering/box.dart#L2402-L2407

Reply:
it's just a graphical effect. It's not interactive and you can't keep it in the transient overscroll state.

Reply:
Is it possible for an overlength title to get clipped or overlap the trailing widget?

Reply:
Very good point. Worth checking what native iOS does when the large title is already horizontally long enough to almost overflow.

Reply:
Hm. Good point.

Reply:
Ok this is what I understand from the behavior:

The trailing widget in large title nav bars lies above the title, therefore it won't overlap horizontally. Because scaling only happens when the height of the nav bar increases, it can't grow large enough vertically to overlap the navigation items (trailing, leading.)

The title may scale up to 1.15, but it is bound to its horizontal margins. Meaning if the text is almost large enough to overflow or if it is truncated, the title text may ever so slightly scale up to the horizontal margins. If it is already exactly at its horizontal margins, it will not scale. I'll update the PR to handle this edge case.

Reply:
@xster @LongCatIsLooong I'm having trouble setting the max-width condition for the scale transformation. My original thought was to have a condition that would check for the width of the large title and ensure it's less than the nav bar builder's `constraints.maxWidth`, but I can't find a way to get the width of the large title text itself. Any ideas?

Reply:
ah doh, sorry I missed this somehow. 

Yes, it's somewhat intentionally complicated because text layout is expensive, so the API is rather low level. There are 2 ways.

The canonical way is to use a RenderBox the way the FittedBox object uses a RenderBox. During the layout phase, it lets its child calculate its size then reacts accordingly in its own paint. 

The slightly less complicated way is to "dry run" it using a TextPainter. There's a bit of that going on in the picker.